esp-config/src/generate/mod.rs

```diff
@@ -45,8 +46,299 @@ impl fmt::Display for Error {
+    Stability::Unstable
+}
+
+/// A default value for a configuration option.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(deny_unknown_fields)]
+pub struct CfgDefaultValue {
+    /// Condition which makes this default value used.
+    /// You can and have to have exactly one active default value.
+    #[serde(rename = "if")]
+    #[serde(default = "true_default")]
+    pub if_: String,
+    /// The default value.
+    pub value: Value,
+}
+
+/// A configuration option.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(deny_unknown_fields)]
+pub struct CfgOption {
+    /// Name of the configuration option
+    pub name: String,
+    /// Description of the configuration option.
+    /// This will be visible in the documentation and in the tooling.
+    pub description: String,
+    /// A condition which specified when this option is active.
+    #[serde(default = "true_default")]
+    pub active: String,
+    /// The default value.
+    /// Exactly one of the items needs to be active at any time.
+    pub default: Vec<CfgDefaultValue>,
+    /// Constraints (Validators) to use.
+    /// If given at most one item is allowed to be active at any time.
+    pub constraints: Option<Vec<CfgConstraint>>,
+    /// A display hint for the value.
+    /// This is meant for tooling and/or documentation.
+    pub display_hint: Option<DisplayHint>,
+    /// The stability guarantees of this option.
+    #[serde(default = "unstable_default")]
+    pub stability: Stability,
+}
+
+/// A conditional constraint / validator for a config option.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(deny_unknown_fields)]
+pub struct CfgConstraint {
+    /// Condition which makes this validator used.
+    #[serde(rename = "if")]
+    #[serde(default = "true_default")]
+    if_: String,
+    /// The validator to be used.
+    #[serde(rename = "type")]
+    type_: Validator,
+}
+
+/// Generate the config from a YAML definition.
+///
+/// The YAML follows the format outlined by [Config].
+///
+/// After deserializing the config and normalizing it, this will call
+/// [generate_config] to finally get the currently active configuration.
+pub fn generate_config_from_yaml_definition(
+    yaml: &str,
+    enable_unstable: bool,
+    emit_md_tables: bool,
+    chip: Option<esp_metadata::Config>,
+) -> Result<HashMap<String, Value>, Error> {
+    let features: Vec<String> = env::vars()
+        .filter(|(k, _)| k.starts_with("CARGO_FEATURE_"))
+        .map(|(k, _)| k)
+        .map(|v| {
+            v.strip_prefix("CARGO_FEATURE_")
+                .unwrap_or_default()
+                .to_string()
+        })
+        .collect();
+
+    let (config, options) = evaluate_yaml_config(yaml, chip, features, false)?;
+
+    let cfg = generate_config(&config.krate, &options, enable_unstable, emit_md_tables);
+
+    if let Some(checks) = config.checks {
+        println!("{:#?}", cfg);
```

Comment:
I'm not sure if you forgot some debug prints, or we want to keep these?

Reply:
Yes - forgot about this, thanks 👍